src/controller.cpp

```diff
@@ -0,0 +1,1230 @@
+        forward_kinematics_matrix_[1][0]*x_alpha + forward_kinematics_matrix_[1][1]*y_alpha;
+
+    return estimated_base_twist;
+}
+
+TricycleSteeringCmdVel Controller::computeTricycleModelInverseKinematics(geometry_msgs::Twist cmd_vel)
+{
+    TricycleSteeringCmdVel steering_cmd_vel;
+    double x_alpha =  inverse_kinematics_matrix_[0][0]*cmd_vel.linear.x
+                    + inverse_kinematics_matrix_[0][1]*cmd_vel.angular.z;
+    double y_alpha =  inverse_kinematics_matrix_[1][0]*cmd_vel.linear.x
+                    + inverse_kinematics_matrix_[1][1]*cmd_vel.angular.z;
+
+    steering_cmd_vel.steering_angle = atan2(y_alpha, x_alpha);
+    steering_cmd_vel.speed = hypot(x_alpha, y_alpha);
+
+    return steering_cmd_vel;
+}
+
+void Controller::setPlan(geometry_msgs::Transform current_tf, geometry_msgs::Twist odom_twist,
+                         const std::vector<geometry_msgs::PoseStamped>& global_plan)
+{
+  ROS_DEBUG("TrackingPidLocalPlanner::setPlan(%d)", (int)global_plan.size());
+  ROS_DEBUG("Plan is defined in frame '%s'", global_plan.at(0).header.frame_id.c_str());
+
+  tf2::Transform current_tf2;
+  tf2::convert(current_tf, current_tf2);
+
+  /* Minimal sanity check */
+  global_plan_tf_.clear();
+  tf2::Transform transform;
+  tf2::fromMsg(global_plan[0].pose, transform);
+  global_plan_tf_.push_back(transform);
+  // For now do not allow repeated points or in-place rotation
+  // To allow that the way the progress is checked and the interpolation is done needs to be changed
+  // Also check if points suddenly go in the opposite direction, this could lead to deadlocks
+  for (int pose_idx = 1; pose_idx < global_plan.size() - 1; ++pose_idx)
+  {
+    auto prev_pose = global_plan[pose_idx - 1].pose;
+    auto pose = global_plan[pose_idx].pose;
+    auto next_pose = global_plan[pose_idx + 1].pose;
+    // Check if the angle of the pose is obtuse, otherwise warn and ignore this pose
+    // We check using Pythagorean theorem: if c*c > (a*a + b*b) it is an obtuse angle and thus we can follow it
+    double a_squared = distSquared(prev_pose, pose);
+    double b_squared = distSquared(pose, next_pose);
+    double c_squared = distSquared(prev_pose, next_pose);
+    if (c_squared > (a_squared + b_squared))
+    {
+      tf2::fromMsg(pose, transform);
+      global_plan_tf_.push_back(transform);
+    }
+    else
+    {
+      ROS_WARN("Pose %d of path is not used since it is not in the expected direction of the path!", pose_idx);
+    }
+  }
+  // Add last pose as we didn't evaluate that one
+  tf2::Transform last_transform;
+  tf2::fromMsg(global_plan.back().pose, last_transform);
+  global_plan_tf_.push_back(last_transform);
+
+  if (!track_base_link_enabled_)
+  {
+    // Add carrot length to plan using goal pose (we assume the last pose contains correct angle)
+    tf2::Transform carrotTF;
+    carrotTF.setOrigin(tf2::Vector3(l_, 0.0, 0.0));
+    global_plan_tf_.push_back(last_transform * carrotTF);
+  }
```

Comment:
> But what if this pose is misaligned with the last path segment?

That would violate the conditions this controller operates in. Although I'm not sure we validate that.

> what if the carrot distance is changed during execution of the plan

Ah yes, that's probably an oversight!

Reply:
This is indeed an issue when following the original path with a carrot, in that case we should only extend the final goal if the end velocity is zero, but of course it can change online, we should have a discussion on what is the best to do.

Reply:
Raised issue #83 for this